js/services.js

```diff
@@ -0,0 +1,166 @@
+          iframe.focus();
+        }
+      }
+      // remove the old one and recreate it.
+      instance.messenger.stop();
+
+      instance.iframe.parentElement.removeChild(instance.iframe);
+      if (instance.config.autoRestart) {
+        this._createService(instance.config);
+      }
+    },
+
+    handleEvent: function(e) {
+      var instance = this._serviceMap[e.target.getAttribute('mozapp')];
+      switch(e.type) {
+        case 'mozbrowsererror':
+          this.handleBrowserError(instance);
+          break;
+        case 'mozbrowserloadend':
+          this.handleBrowserLoaded(instance);
+          break;
+      }
+    },
+
+    _processCreateQueue: function() {
+      if (this._clonedConfig.length === 0) {
+        this.send({'log': 'all services created'});
+        delete this._clonedConfig;
+        this._dispatchReady();
+        return;
+      }
+      this._createService(this._clonedConfig.pop());
+
+    },
+
+    _createService: function(s) {
+      if (!s) {
+        return;
+      }
+      // create iframe
+      var iframe = document.createElement('iframe');
+      iframe.setAttribute('mozbrowser', 'true');
+      if (s.type === 'app') {
+        iframe.setAttribute('remote', 'true');
+      }
+      iframe.setAttribute('mozapp', s.manifestURL);
+      iframe.setAttribute('mozallowfullscreen', 'true');
+      iframe.setAttribute('origin', s.origin);
+      iframe.src = s.origin + s.launchURL;
+      iframe.hidden = true;
+      iframe.addEventListener('mozbrowsererror', this);
+      iframe.addEventListener('mozbrowserloadend', this);
+
+      // register
+      this._serviceMap[s.manifestURL] = {
+        'config': s,
+        'iframe': iframe
+      };
+      // put on DOM
+      if (s.type === 'window' || s.type === 'app') {
+        this._windowContainer.appendChild(iframe);
+      } else {
+        this._serviceContainer.appendChild(iframe);
+      }
+
+      this.send({'log': 'service [' + s.manifestURL + '] created'});
+      return this._serviceMap[s.manifestURL];
+    },
+
+    send: function(data, targetOrigin) {
+      for (var url in this._serviceMap) {
+        if (targetOrigin &&
+            targetOrigin !== this._serviceMap[url].messenger.targetOrigin) {
+          continue;
+        }
+        // someone who doesn't have message is loading its resource.
+        if (this._serviceMap[url].messenger) {
+          this._serviceMap[url].messenger.send(data);
+        }
+      }
+    },
+
+    start: function() {
+      this._serviceMap = {};
+      this._serviceContainer = document.getElementById('service-section');
+      this._windowContainer = document.getElementById('window-section');
+      // We need to be sure to get the focus in order to wake up the screen
+      // if the phone goes to sleep before any user interaction.
+      // Apparently it works because no other window
+      // has the focus at this point.
+      window.focus();
+
+      this._clonedConfig = [].concat(ServiceConfig);
```

Comment:
I don't have a good name now. And initialLaunchList is improper, too. I will keep it for renaming issue.